helpers.go

```diff
@@ -267,29 +268,111 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		return offers[0]
 	}
 
-	for _, offer := range offers {
-		if len(offer) == 0 {
-			continue
+	type acceptedType struct {
+		spec        string
+		quality     float64
+		specificity int
+		order       int
+	}
+
+	// Parse header and get accepted types with their quality and specificity
+	// See: https://www.rfc-editor.org/rfc/rfc9110#name-content-negotiation-fields
+	spec, commaPos, order := "", 0, 0
+	acceptedTypes := make([]acceptedType, 0, 20)
+	for len(header) > 0 {
+		order++
+
+		// Skip spaces
+		header = utils.TrimLeft(header, ' ')
+
+		// Get spec
+		commaPos = strings.IndexByte(header, ',')
+		if commaPos != -1 {
+			spec = utils.Trim(header[:commaPos], ' ')
+		} else {
+			spec = utils.TrimLeft(header, ' ')
+		}
+
+		// Get quality
+		quality := 1.0
+		if factorSign := strings.IndexByte(spec, ';'); factorSign != -1 {
+			factor := utils.Trim(spec[factorSign+1:], ' ')
+			if strings.HasPrefix(factor, "q=") {
+				if q, err := strconv.ParseFloat(factor[2:], 64); err == nil {
+					quality = q
+				}
+			}
+			spec = spec[:factorSign]
 		}
-		spec, commaPos := "", 0
-		for len(header) > 0 && commaPos != -1 {
-			commaPos = strings.IndexByte(header, ',')
+
+		// Skip if quality is 0.0
+		// See: https://www.rfc-editor.org/rfc/rfc9110#quality.values
+		if quality == 0.0 {
 			if commaPos != -1 {
-				spec = utils.Trim(header[:commaPos], ' ')
+				header = header[commaPos+1:]
 			} else {
-				spec = utils.TrimLeft(header, ' ')
-			}
-			if factorSign := strings.IndexByte(spec, ';'); factorSign != -1 {
-				spec = spec[:factorSign]
+				break
 			}
+			continue
+		}
 
-			// isAccepted if the current offer is accepted
-			if isAccepted(spec, offer) {
-				return offer
+		// Get specificity
+		specificity := 0
+		// check for wildcard this could be a mime */* or a wildcard character *
+		if spec == "*/*" || spec == "*" {
+			specificity = 1
+		} else if strings.HasSuffix(spec, "/*") {
+			specificity = 2
+		} else if strings.IndexByte(spec, '/') != -1 {
+			specificity = 3
+		} else {
+			specificity = 4
+		}
+
+		// Add to accepted types
+		acceptedTypes = append(acceptedTypes, acceptedType{spec, quality, specificity, order})
+
+		// Next
+		if commaPos != -1 {
+			header = header[commaPos+1:]
+		} else {
+			break
+		}
+	}
+
+	// Sort accepted types by quality and specificity, preserving order of equal elements
+	if len(acceptedTypes) > 1 {
```

Comment:
can we create a function from this to slim down the function body?

i know in golang it is common practice to have big functions and files, but i am a friend of the clean code concept with small parts

If it does not add allocation, then please convert to separat function
(same with the other parts, when possible)

Reply:
done